deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/beans/OperationResponse.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.testgrid.deployment.tinkerer.beans;
+
+/**
+ * This class holds operation response data.
+ *
+ * @since 1.0.0
+ */
+public class OperationResponse extends Operation {
```

Comment:
I would expect that a sub-class of Operation does a specialized operation. For example, something like ExecuteShellOperation, LogStreamOperation, FileArchiveOperation etc. 

The purpose of OperationResponse is a bit unclear. :)

Reply:
Operation request contains the request coming from the user while Operation response contains the response for operations. Both are forms of abstract operation and shares same data and functionalities while operation response holds the response for the operation and operation request holds the original request from the user.
In here we don't need to have different sub classes to model different operations as long as operation it self supported to cater different types of operations with operation code attribute.

Reply:
This is an interesting discussion. Let's continue this outside this PR. :)

Reply:
I think the Operations need to be given first-class status. It should not hidden inside somewhere as an enum. An operation need to be a concrete class with an #execute method that knows how to execute that operation. I think execute-shell, ping, log-archive are all operations that need to have their on #execute implementations.

At the same time, the request/response analogy that you bring up is quite same as [HttpServletRequest](https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html) and [HttpServletResponse](https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletResponse.html). I do not see the parent-child hierarchy you mention in there.

Going further, shall we think how we improve this? I think we can model this similar to how Axis2 cluster messages work. In Axis2, Cluster messages is a way to send instructions and some state to other nodes. So, how about a class-hierarchy like this that I based on your impl and axis2 cluster message design:

```java
interface Operatable {
    String execute();
}

class OperationResponse {
  String uuid;
  public String getUUID() {}
}

class ExecuteShellOperation implements Operatable {
  public String execute() {}
}
class PINGOperation implements Operatable {
  public String execute() {}
}

class ExecuteShellOperationResponse extends OperationResponse {}
class PINGOperationResponse extends OperationResponse {}

```

Then, you can send a message to an agent like follows: 
```java
class HeartBeatChecker {
  public void sendHeartBeat(SessionManager sessionManager, String agentId) {
    Operatable op = new PINGOperation();
    String uuid = sessionManager.sendOperation(op);
    OperationResponse opResp = sessionManager.getResponse(uuid); // blocking as in j.u.c.Future.get(). We can provide an async method as well.
  }
}
```

We can obviously fine-tune this. WDYT about the overall design?